[Bug]: system npm upgrade can leave running gateway in transient half-swapped package/plugin state

### Bug type

Behavior bug (incorrect output/state without crash)

### Beta release blocker

No

### Summary

During a live Linux system-global upgrade from `2026.5.16-beta.3` to `2026.5.18`, the running gateway briefly loaded a half-swapped package/plugin state and logged missing bundled modules, including `@openclaw/codex` missing `./string-coerce-LndEvhRk.js`.

### Steps to reproduce

1. Start from a root-owned system-global npm install where the user gateway service runs from `/usr/lib/node_modules/openclaw`.
2. Run `openclaw update --channel stable --yes` as the non-root gateway user.
3. Observe the update fail at the global install stage with `EACCES`, then restart the existing gateway.
4. Run the documented/manual system package recovery path with the system npm binary: `sudo -n /usr/bin/npm install -g openclaw@latest` while the gateway service is still running.
5. Observe the gateway logs during the npm package replacement window.
6. Complete the install, run `HOME=/home/ubuntu openclaw gateway install --force`, restart the gateway, and observe that the missing-module condition clears.

### Expected behavior

The update/recovery path should not leave the running gateway loading from a package tree while npm is replacing bundled files.

For a system-global install requiring sudo, OpenClaw should either keep the gateway stopped through the package replacement, fail closed with an explicit instruction to stop the gateway before manual npm install, or otherwise avoid transient plugin/core module resolution failures during package replacement.

### Actual behavior

The gateway attempted to continue loading runtime/plugin code during the package replacement window and logged missing modules from both the user-managed plugin tree and the global core dist tree.

After the install completed, `openclaw doctor --fix`, plugin registry refresh, and a clean gateway restart cleared the condition. The failure was transient rather than persistent.

### OpenClaw version

Upgrade path: `OpenClaw 2026.5.16-beta.3 (d08cbf7)` -> `OpenClaw 2026.5.18 (50a2481)`

### Operating system

Ubuntu 24.04.3 LTS, Linux `6.17.0-1011-oracle`, `aarch64`

### Install method

System-global npm install, root-owned prefix:

```text
node=v22.22.0
npm=10.9.4
npm root -g=/usr/lib/node_modules
install root=/usr/lib/node_modules/openclaw
service=/home/ubuntu/.config/systemd/user/openclaw-gateway.service
ExecStart=/usr/bin/node /usr/lib/node_modules/openclaw/dist/index.js gateway --port 18789
```

### Model

`openai/gpt-5.5`

### Provider / routing chain

OpenClaw gateway -> OpenAI Codex/ChatGPT auth, effective Codex harness for agent turns.

### Additional provider/model setup details

Enabled plugins after recovery: `active-memory`, `anthropic`, `file-transfer`, `google`, `memory-core`, `memory-wiki`, `ollama`, `openai`, `telegram`, `tokenjuice`, `codex`.

`codex` plugin after recovery:

```text
source=/home/ubuntu/.openclaw/npm/node_modules/@openclaw/codex/dist/index.js
rootDir=/home/ubuntu/.openclaw/npm/node_modules/@openclaw/codex
version=2026.5.6
status=loaded
dependencyStatus.requiredInstalled=true
```

### Logs, screenshots, and evidence

```shell
# Initial OpenClaw updater attempt as the gateway user
Updating OpenClaw...

Stopping managed gateway service before package update...
Stopped systemd service: openclaw-gateway.service

Update Result: ERROR
  Root: /usr/lib/node_modules/openclaw
  Reason: global install stage
  Before: 2026.5.16-beta.3
  After: 2026.5.16-beta.3

Steps:
  ✗ global install stage (0ms)
      EACCES: permission denied, mkdtemp '/usr/lib/node_modules/.openclaw-update-stage-S61KvY'

Recovery hints:
  - Detected permission failure (EACCES). Re-run with a writable global prefix or sudo (for system-managed Node installs).
  - Example: npm config set prefix ~/.local && npm i -g openclaw@latest

Restarted systemd service: openclaw-gateway.service
Restarted managed gateway service after failed update.
```

```shell
# During the following system npm install window
May 18 19:39:34 polymarket-mc node[1178137]: [plugins] codex failed to load from /home/ubuntu/.openclaw/npm/node_modules/@openclaw/codex/dist/index.js: Error: Cannot find module './string-coerce-LndEvhRk.js'
May 18 19:39:34 polymarket-mc node[1178137]: [plugins] 1 plugin(s) failed to initialize (load: codex). Run 'openclaw plugins list' for details.
May 18 19:39:34 polymarket-mc node[1178137]: [gateway] [plugins] failed to load plugin: Error: Cannot find module './string-coerce-LndEvhRk.js'
May 18 19:39:34 polymarket-mc node[1178137]: [ws] failed to load ws message handler conn=11815436-d88e-452a-b82d-9728df47c114: Cannot find module '/usr/lib/node_modules/openclaw/dist/message-handler-D5F65c4n.js' imported from /usr/lib/node_modules/openclaw/dist/server-ws-runtime-DPe-iy_y.js
May 18 19:39:34 polymarket-mc node[1178137]: [openclaw] The CLI command failed.
May 18 19:39:34 polymarket-mc systemd[1245]: openclaw-gateway.service: Failed with result 'exit-code'.
May 18 19:39:40 polymarket-mc node[1178688]: Error: Cannot find module '/usr/lib/node_modules/openclaw/dist/index.js'
May 18 19:39:40 polymarket-mc systemd[1245]: openclaw-gateway.service: Failed with result 'exit-code'.
```

```shell
# Recovery evidence after completed install + doctor --fix + restart
OpenClaw 2026.5.18 (50a2481)
/usr/bin/openclaw -> /usr/lib/node_modules/openclaw/openclaw.mjs
package_root=/usr/lib/node_modules/openclaw
readyz={"ready":true,"failing":[]}
plugins list: count=92, enabled=11, errors=[]
openclaw doctor --deep --lint --json: findingCount=0
```

### Impact and severity

Affected: Linux/systemd users on root-owned system-global npm installs who recover from updater `EACCES` using manual `sudo npm install -g openclaw@latest` while the gateway is running.

Severity: Medium. This creates a transient gateway outage and confusing plugin/core missing-module diagnostics during the upgrade window.

Frequency: Observed once on a live VPS during `2026.5.16-beta.3` -> `2026.5.18`. NOT_ENOUGH_INFO for broader reproducibility.

Consequence: Gateway can crash/restart during upgrade and report plugin failures that are not persistent after the install finishes.

### Additional information

This did not reproduce as a persistent `@openclaw/codex` install failure. After the completed install, `doctor --fix`, plugin registry refresh, and clean restart, `codex` loaded successfully and plugin list reported zero errors.

This may be adjacent to older upgrade/cutover state issues such as #76865 and plugin sync reports such as #78892, but the observed symptom here was a transient half-swapped package state during manual system npm recovery after an updater `EACCES`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,6 +50,7 @@ Docs: https://docs.openclaw.ai
 
 - Memory/search: scan the JS-side fallback vector path (used when the sqlite-vec index is unavailable or has a mismatched dimension) in bounded rowid batches and yield to the event loop between batches so large chunk tables can no longer pin the Node.js main thread for multi-second windows. Also keeps the SQL prepared statement rooted in a local so node:sqlite cannot finalize it mid-scan under heap pressure. Fixes #81172. Thanks @dev23xyz-oss.
 - CLI/update: bypass npm freshness filters consistently during managed package and plugin installs so freshly published release plugins remain installable. Thanks @jalehman.
+- CLI/update: guide root-owned npm install EACCES recovery by stopping the managed Gateway before manual package replacement, then reinstalling and restarting the service. Fixes #83747. (#83757) Thanks @brokemac79.
 - Agents/subagents: keep collect-mode announce queues batching unresolved-origin items with compatible same-route messages and resume collection after a true cross-channel drain when a later compatible batch remains. Fixes #83577.
 - Providers/Anthropic: preserve native image input for current Claude model rows when stale local catalog data marks them text-only. (#83756) Thanks @TurboTheTurtle.
 - Control UI: render live tool progress from session-scoped `session.tool` Gateway events so externally started runs show their tool cards in the active session. (#83734) Thanks @TurboTheTurtle.

--- a/docs/install/updating.md
+++ b/docs/install/updating.md
@@ -124,7 +124,8 @@ Then verify the service:
 openclaw --version
 curl -fsS http://127.0.0.1:18789/readyz
 openclaw plugins list --json
-openclaw doctor --deep --lint --json
+openclaw gateway status --deep --json
+openclaw doctor --lint --json
 ```
 
 When `openclaw update` manages a global npm install, it installs the target into

--- a/docs/install/updating.md
+++ b/docs/install/updating.md
@@ -98,10 +98,34 @@ npm i -g openclaw@latest
 ```
 
 Prefer `openclaw update` for supervised installs because it can coordinate the
-package swap with the running Gateway service. If you update manually while a
-managed Gateway is running, restart the Gateway immediately after the package
-manager finishes so the old process does not keep serving from replaced package
-files.
+package swap with the running Gateway service. If you update manually on a
+supervised install, stop the managed Gateway before the package manager starts.
+Package managers replace files in place, and a running Gateway can otherwise try
+to load core or plugin files while the package tree is temporarily half-swapped.
+Restart the Gateway after the package manager finishes so the service picks up
+the new install.
+
+For a root-owned Linux system-global install, if `openclaw update` fails with
+`EACCES` and you recover with system npm, keep the Gateway stopped through the
+manual package replacement. Use the same `openclaw` profile flags or environment
+you normally use for that Gateway. Replace `/usr/bin/npm` with the system npm
+that owns the root-owned global prefix on your host:
+
+```bash
+openclaw gateway stop
+sudo /usr/bin/npm i -g openclaw@latest
+openclaw gateway install --force
+openclaw gateway restart
+```
+
+Then verify the service:
+
+```bash
+openclaw --version
+curl -fsS http://127.0.0.1:18789/readyz
+openclaw plugins list --json
+openclaw doctor --deep --lint --json
+```
 
 When `openclaw update` manages a global npm install, it installs the target into
 a temporary npm prefix first, verifies the packaged `dist` inventory, then swaps

--- a/src/cli/update-cli/progress.test.ts
+++ b/src/cli/update-cli/progress.test.ts
@@ -64,6 +64,7 @@ describe("inferUpdateFailureHints", () => {
     const hints = inferUpdateFailureHints(result);
     expect(hints.join("\n")).toContain("EACCES");
     expect(hints.join("\n")).toContain("npm config set prefix ~/.local");
+    expect(hints.join("\n")).toContain("stop the Gateway first");
   });
 
   it("returns EACCES hint for staged package permission failures", () => {
@@ -74,6 +75,9 @@ describe("inferUpdateFailureHints", () => {
     const hints = inferUpdateFailureHints(result);
     expect(hints.join("\n")).toContain("EACCES");
     expect(hints.join("\n")).toContain("npm config set prefix ~/.local");
+    expect(hints.join("\n")).toContain("<system-npm>");
+    expect(hints.join("\n")).toContain("gateway install --force");
+    expect(hints.join("\n")).toContain("gateway restart");
   });
 
   it("returns native optional dependency hint for node-gyp failures", () => {

--- a/src/cli/update-cli/progress.ts
+++ b/src/cli/update-cli/progress.ts
@@ -85,7 +85,13 @@ export function inferUpdateFailureHints(result: UpdateRunResult): string[] {
     hints.push(
       "Detected permission failure (EACCES). Re-run with a writable global prefix or sudo (for system-managed Node installs).",
     );
+    hints.push(
+      "If you recover with sudo/manual package install on a managed Gateway, stop the Gateway first so it does not load files while the package tree is being replaced.",
+    );
     hints.push("Example: npm config set prefix ~/.local && npm i -g openclaw@latest");
+    hints.push(
+      "System install outline: openclaw gateway stop -> sudo <system-npm> i -g openclaw@latest -> openclaw gateway install --force -> openclaw gateway restart.",
+    );
   }
 
   if (

